### PR TITLE
prime_factors & nth_prime

### LINKED ISSUE #205: Add Practice Exercise: `nth-prime`
Problem specification:
https://github.com/exercism/problem-specifications/tree/main/exercises/nth-prime
    
TODO:
    
- generate the exercise with `./bin/fetch-configlet && ./bin/configlet create --practice-exercise nth-prime`
- check if any of the existing tracks (examples below) implement the exercise and you can reuse the implementation:
    - Go: https://github.com/exercism/go/tree/main/exercises/practice
    - Rust: https://github.com/exercism/rust/tree/main/exercises/practice
    - Python: https://github.com/exercism/python/tree/main/exercises/practice
- use exercise's _canonical-data.json_ to see which test cases to implement:
    https://github.com/exercism/problem-specifications/blob/main/exercises/nth-prime/canonical-data.json

### LINKED ISSUE #215: Add Practice Exercise: `prime-factors`
Problem specification:
https://github.com/exercism/problem-specifications/tree/main/exercises/prime-factors
    
TODO:
    
- generate the exercise with `./bin/fetch-configlet && ./bin/configlet create --practice-exercise prime-factors`
- check if any of the existing tracks (examples below) implement the exercise and you can reuse the implementation:
    - Go: https://github.com/exercism/go/tree/main/exercises/practice
    - Rust: https://github.com/exercism/rust/tree/main/exercises/practice
    - Python: https://github.com/exercism/python/tree/main/exercises/practice
- use exercise's _canonical-data.json_ to see which test cases to implement:
    https://github.com/exercism/problem-specifications/blob/main/exercises/prime-factors/canonical-data.json

### PR DESCRIPTION
**Pull request by Falilah**
Closes #205 
Closes #215

### CODE REVIEW

**Comment by 0xNeshi** [comment]
Hi, thanks for the PR! Will be taking a look later in the day

File: exercises/practice/nth-prime/tests/nth_prime.cairo [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#[test]
+#[ignore]
+fn big_prime() {
+    let output = nth(500);
+    let expected = 3571;
+    assert_eq!(output, expected);
+}
+
+#[test]
+#[should_panic(expected: ("there is no zeroth prime",))]
```
Comment (0xNeshi) [inline exercises/practice/nth-prime/tests/nth_prime.cairo:35]:
```suggestion
#[ignore]
#[should_panic(expected: ("there is no zeroth prime",))]
```
forgot to ignore the test

File: exercises/practice/prime-factors/.meta/example.cairo [hunk 2]
```diff
@@ -0,0 +1,20 @@
+pub fn factors(n: u64) -> Array<u64> {
+    let mut all_prime = ArrayTrait::new();
+
+    let mut num: u64 = n;
+    loop {
+        for i in 2..(num + 1) {
```
Comment (0xNeshi) [inline exercises/practice/prime-factors/.meta/example.cairo:6]: Although this works, it seems kind of redundant to continuously recreate a for loop + range that goes to `num + 1`, but then almost never reach it. 
Example solutions are there to showcase an idiomatic way to solve the problem for a given language, so we need to pay attention to clean code.
There is a way to write this function without depending on breaks and never-to-be-reached ranges, try to rewrite it in such a way. You can use [Rust's](https://github.com/exercism/rust/blob/main/exercises/practice/prime-factors/.meta/example.rs) implementation as inspiration.

**Review by 0xNeshi — CHANGES_REQUESTED** [review]
Very good job! Left some relatively minor comments, we can merge immediately after those are addressed

File: exercises/practice/nth-prime/tests/nth_prime.cairo [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#[ignore]
+fn sixth_prime() {
+    let output = nth(6);
+    let expected = 13;
+    assert_eq!(output, expected);
+}
+
+#[test]
+#[ignore]
+fn big_prime() {
```
[the remark this answers is not in the document]
Comment (Falilah) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: making it compliant with canonical data which is 10001 is throwing `out of gas` error when tested, that is why I reduced the value to 500, what do you think we do?

[the remark this answers is not in the document]
Comment (0xNeshi) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: Hmm, try rerunning the test with an [updated implementation](https://github.com/exercism/cairo/pull/282#discussion_r1769008529). Note that if you reuse Rust's implementation, you can increase `possible` by 2 to look only at odd numbers, thus improving efficiency potentially by 2x. Let me know if that works

[the remark this answers is not in the document]
Comment (0xNeshi) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: Let's leave the test in with a smaller value than in canonical data, but let's also make 2 changes:
1. increase the value to `4001` and the expected value to `37831` 
2. add a comment to [tests.toml below this test description](https://github.com/exercism/cairo/pull/282/files#diff-14a49407b23723cb62218e8d3afd6491c34a4d2d81dd7b434740bf80b8648e51R22) stating: `comment = "decreased the input value to fit into Cairo's runtime limitation"

[the remark this answers is not in the document]
Comment (Falilah) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: I increased the value to 4001, it is  passing locally but the workflow timed out with `out of gas`. I then bring it down to  500 which was what i had before, the  run is still failing with `out of gas`. I have left it in the suggested value which is 4001, kindly help look into the job being run.

[the remark this answers is not in the document]
Comment (0xNeshi) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: The problem is not with `nth_prime`, but with `prime-factors` :D But there is an `out of gas` issue there.
You can solve the issue by the following optimizations:
- first handle all divisions by `2`, which is the only even prime number
- then starting from `3`, check for prime factors, increasing the possible prime not by 1, but by 2

```rust
pub fn factors(n: u64) -> Array<u64> {
    let mut val = n;
    let mut out: Array<u64> = ArrayTrait::new();

    if val < 2 {
        return out;
    }

    while val % 2 == 0 {
        out.append(2);
        val /= 2;
    };

    let mut possible = 3;
    while val != 1 {
        while val % possible == 0 {
            out.append(possible);
            val /= possible;
        };
        possible += 2;
    };

    out
}
```

[the remark this answers is not in the document]
Comment (Falilah) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: oooh my bad too, I had the prime numbers all passing before I started the nth, but I guess I did not pay enough attention to the details od the workflow

[the remark this answers is not in the document]
Comment (Falilah) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: seems we are good now. I have adjusted the solution according to yours.
let me know if there is any changes to be made

[the remark this answers is not in the document]
Comment (0xNeshi) [reply exercises/practice/nth-prime/tests/nth_prime.cairo:28]: All good now

**Review by 0xNeshi — APPROVED** [review]
Very good work!

**Pull request merged**